ingest/sqs/ingester.go

```diff
@@ -1,55 +1,281 @@
-	poolsIngester     domain.AtomicIngester
-	chainInfoIngester domain.AtomicIngester
+	poolsIngester domain.BlockProcessor
+	grpcConn      *grpc.ClientConn
+	appCodec      codec.Codec
+	mx            *sync.Mutex
+	lastBlockTime time.Time
+
+	isProcessingBlock atomic.Bool
 }
 
 // NewSidecarQueryServerIngester creates a new sidecar query server ingester.
 // poolsRepository is the storage for pools.
 // gammKeeper is the keeper for Gamm pools.
-func NewSidecarQueryServerIngester(poolsIngester, chainInfoIngester domain.AtomicIngester, txManager repository.TxManager) ingest.Ingester {
+func NewSidecarQueryServerIngester(poolsIngester domain.BlockProcessor, appCodec codec.Codec) ingest.Ingester {
 	return &sqsIngester{
-		txManager:         txManager,
-		chainInfoIngester: chainInfoIngester,
-		poolsIngester:     poolsIngester,
+		poolsIngester: poolsIngester,
+		appCodec:      appCodec,
+
+		mx:            &sync.Mutex{},
+		lastBlockTime: time.Unix(0, 0),
+
+		isProcessingBlock: atomic.Bool{},
 	}
 }
 
+type IngestProcessBlockArgs struct {
+	Pools []sqsdomain.PoolI
+}
+
 // ProcessBlock implements ingest.Ingester.
-func (i *sqsIngester) ProcessBlock(ctx sdk.Context) error {
-	// Start atomic transaction
-	tx := i.txManager.StartTx()
+func (i *sqsIngester) ProcessBlock(ctx sdk.Context) (err error) {
+	// This mechanism ensures that we do not process the next block while syncing.
+	if i.isProcessingBlock.Load() {
+		return nil
+	}
+
+	// Set processing block to true
+	i.isProcessingBlock.Store(true)
+
+	// Note, that this has to be async as SQS might query the node for CW contract estimates.
+	// As a result, SQS data ingestion must not be blocked by this.
+	go func() {
+		defer func() {
+			// Set processing block to false
+			i.isProcessingBlock.Store(false)
+
+			// Measure block processing time
+			telemetry.MeasureSince(i.lastBlockTime, "sqs_block_process_time")
+
+			// Update last block time for syncing detection
+			i.lastBlockTime = time.Now()
+
+			var panicErr any
+			if panicErr = recover(); panicErr != nil {
+				telemetry.IncrCounterWithLabels([]string{"sqs_block_process_error"}, 1, []metrics.Label{
+					{
+						Name:  "err",
+						Value: fmt.Sprintf("panic: %v", panicErr),
+					},
+				})
+			}
+
+			if err != nil {
+				telemetry.IncrCounterWithLabels([]string{"sqs_block_process_error"}, 1, []metrics.Label{
+					{
+						Name:  "err",
+						Value: err.Error(),
+					},
+				})
+			}
+
+			// If error or panic occurred, close the grpc connection
+			// so that it can be attempted to be re-established in the next block processing
+			if (panicErr != nil || err != nil) && i.grpcConn != nil {
+				i.grpcConn.Close()
+				i.grpcConn = nil
+			}
+		}()
+
+		// syncing detecting
+		timeSinceLastBlock := time.Since(i.lastBlockTime)
+		if timeSinceLastBlock < syncingThreshold {
+			ctx.Logger().Info("syncing detected, skipping block processing", "height", ctx.BlockHeight(), "time_since_last_block", timeSinceLastBlock)
```

Comment:
curious, how does the ctx.Logger work in sqs? Does it get logged together in the daemon's main log? (Asking since I remember there was a seperate log for sqs)

Reply:
Since this utilizes the default context logger, this specific one would go into the main log.

The web server has its own logger hat writes to file